src/mem/cache/prefetch/indirect_memory_v2.hh

```diff
@@ -0,0 +1,662 @@
+                     * the index into the IPD_table
+                     */
+                    std::optional<Addr> active_IP;
+
+                    /*
+                     * List of supported shift values
+                     */
+                    const std::vector<int> shift_values;
+
+                    /*
+                     * Number of misses to track after an index access
+                     */
+                    const uint64_t num_miss_after_index;
+
+                public:
+                    /*
+                     * This type represents the output of the IPD. The goal of
+                     * the IPD is to compute the indirect access parameters for
+                     * a given IP (which points to the load instruction
+                     * responsible for accessing the indexing array B), i.e a
+                     * (base address, shift) pair. Since these parameters need
+                     * to be stored in the indirect table (which is indexed
+                     * using the IP responsible for reading from the indexing
+                     * array, B), IPD must also return the IP for which it
+                     * computed the (base address, shift) pair
+                     */
+                    typedef std::tuple<Addr, Addr, uint64_t>
+                        IndirectParameters;
+
+                    IndirectPatternDetector(const std::size_t size,
+                            const uint64_t num_miss,
+                            const std::vector<int>&& shift_vals):
+                        IPD_table_size(size), active_IP({}),
+                        shift_values(std::move(shift_vals)),
+                        num_miss_after_index(num_miss)
+                        {}
+
+                    /*
+                     * Set the index1 and index2 fields for the entry
+                     * corresponding to IP
+                     */
+                    void setIndex(const Addr IP, const int64_t index);
+
+                    /*
+                     * Discard the most recently observed index
+                     */
+                    void discardIndex();
+
+                    /*
+                     * Record a miss and compute the indirect access parameters
+                     * miss_addr is the cache access address which missed and
+                     * IMP assumes that the miss_addr corresponds to an access
+                     * to the indirect array, A
+                     */
+                    std::optional<IndirectParameters> recordMiss(
+                            const Addr miss_addr);
+            };
+        } // namespace IMPv2Internals
+
+        class IMP : public Queued
+        {
+            private:
+                IMPv2Internals::PrefetchTable prefetch_table;
+                IMPv2Internals::IndirectPatternDetector ipd;
+
+                /*
+                 * The degree for streaming and indirect prefetches
```

Comment:
nit: ´degree´ does not give much information for people that not very familiar with prefetchers. Do you mind changing the documentation to something more informative? It still makes sense to call the variable a degree, but I like how the documentation was declared in old IMP:  /** Number of prefetches generated when using the streaming prefetcher */